### ISSUE TITLE
use_vmac problems with getting broadcast address of new interface

### ISSUE THREAD

**Issue by tciops** [acassen/keepalived#1111]
I've faced an issue with hazelcast that is running on a machine which has VIP.
Hazelcast was not able to start and in log I found piece which says:
```
java.net.SocketException: Cannot assign requested address (ioctl SIOCGIFBRDADDR failed)
        at java.net.NetworkInterface.getAll(Native Method)
        at java.net.NetworkInterface.getNetworkInterfaces(NetworkInterface.java:343)
        at com.hazelcast.instance.DefaultAddressPicker.pickMatchingAddress(DefaultAddressPicker.java:322)
        at com.hazelcast.instance.DefaultAddressPicker.pickInterfaceAddress(DefaultAddressPicker.java:200)
        at com.hazelcast.instance.DefaultAddressPicker.pickAddress(DefaultAddressPicker.java:164)
        at com.hazelcast.instance.DefaultAddressPicker.getPublicAddressByPortSearch(DefaultAddressPicker.java:104)
        at com.hazelcast.instance.DefaultAddressPicker.pickAddress(DefaultAddressPicker.java:74)
```

I checked the code of NetworkInterface.getAll and found that for all interfaces that are found it tries to get broadcast address with ioctl.
Knowing this I've added brd to virual_address but to no avail.

I'm not entirely sure that this is a keepalived issue (maybe linux kernel?) but I'd rather start here first.

My keepalived conf:

```
vrrp_instance app {
    state MASTER
    interface bond0
    virtual_router_id 52
    # for backup node change this to 100
    priority 101
    advert_int 1
    use_vmac vrrpint
    vmac_xmit_base
    preempt
    authentication {
        auth_type PASS
        auth_pass 1111
    }
    virtual_ipaddress {
        192.168.100.25/24 brd 192.168.100.255 label bond0:vip
    }
    notify_master "/etc/keepalived/vmac_tweak.sh"
    notify_backup  "/etc/keepalived/vmac_tweak.sh"
}
```
vmac_tweak works exactly as described in [NOTE_vrrp_vmac.txt](https://github.com/acassen/keepalived/blob/master/doc/NOTE_vrrp_vmac.txt)

I've used this piece of C code to test if various settings in keepalived will let me get this working - but to no avail.
Part of this code is a copy from [NetworkInterface](https://github.com/samskivert/ikvm-openjdk/blob/master/jdk/src/share/classes/java/net/NetworkInterface.java) implementation JDK
```
#include <netinet/in.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <arpa/inet.h>
#include <net/if.h>
#include <net/if_arp.h>


#ifdef __linux__
#include <sys/ioctl.h>
#include <bits/ioctls.h>
#include <sys/utsname.h>
#include <stdio.h>
#endif




main ( int argc, char **argv ) {

        // sock = openSocket(env, AF_INET);
        // static int openSocket(JNIEnv *env, int proto){
        // sock = JVM_Socket(AF_INET, SOCK_DGRAM, 0)

        // JVM_Socket(jint domain, jint type, jint protocol)
        // result = socket(domain, type, protocol);
        int sock  = socket(AF_INET, SOCK_DGRAM, 0);

        struct ifreq if2;

        memset((char *) &if2, 0, sizeof(if2));
        strcpy(if2.ifr_name, argv[1]);

        if (ioctl(sock, SIOCGIFFLAGS, (char *)&if2)  < 0) {
                printf("IOCTL  SIOCGIFFLAGS failed %s",argv[1]);
                return 0;
        }

        if (if2.ifr_flags & IFF_BROADCAST) {
                /* It does, let's retrieve it*/
                if (ioctl(sock, SIOCGIFBRDADDR, (char *)&if2) < 0) {
                        printf("IOCTL SIOCGIFBRDADDR failed %s",argv[1]);
                        return 0;
                }
        }
        

        return 0;

}
```
When I compile it and run with interface name as an argument it will always give me `IOCTL SIOCGIFBRDADDR failed` for virtual interface.

Basically the workaround was to start the app with hazecast and just then start keepalived.
Again, hope this is the right place for this bug.
If it is a 3rd party software - then sorry for taking your time.

**Comment by pqarmitage** [comment]
It appears that the problem is caused by having the label on the address. The address doesn't get considered as a candidate for returning the broadcast address when it has a label. If the virtual_ipaddress is configured without the label then the error doesn't occur.

If you configure an additional IP address within the virtual_ipaddress block, so for example you have:
```
    virtual_ipaddress {
        192.168.100.25/24 brd 192.168.100.255 label bond0:vip
        172.16.0.1/24
    }
```
then the problem should be resolved.

This certainly isn't a keepalived issue. It looks to me it is more likely an issue with Hazelcast, although it might be an issue with java.net.NetworkInterface.getAll or java.net.NetworkInterface.getNetworkInterfaces. It is quite reasonable to have an interface without an address, or an address with a label, and that shouldn't cause an exception in java.net.NetworkInterface.getNetworkInterfaces unless the exception is expected to be caught.
Specifying a broadcast address is not necessary since the SIOCGIFBRDADDR ioctl will return 0.0.0.0 if there is no broadcast address.
